### PR TITLE
Amalgamate statustexts using chunk number and id

### PR DESCRIPTION
**Pull request by peterbarker**
https://github.com/ArduPilot/mavlink/pull/118 <- mavlink PR
https://github.com/ArduPilot/MAVProxy/pull/718 <- MAVProxy PR

`chunk_num` will become `seq`

### CODE REVIEW

**Comment by peterbarker** [comment]
```
2019-12-12 10:42:10.27: STATUSTEXT {severity : 6, text : 266, id : 0, chunk_num : 0} 2019-12-12 10:42:10.27: STATUSTEXT {severity : 6, text : 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01, id : 217, chunk_num : 0} 2019-12-12 10:42:10.28: STATUSTEXT {severity : 6, text : 23456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123, id : 217, chunk_num : 1} 2019-12-12 10:42:10.28: STATUSTEXT {severity : 6, text : 456789ABCDEF0123456789ABCDEF0123456789ABCDEF012345, id : 217, chunk_num : 2} 2019-12-12 10:42:10.28: STATUSTEXT {severity : 6, text : 6789ABCDEF0123456789ABCDEF0123456789ABCDEF01234567, id : 217, chunk_num : 3} 2019-12-12 10:42:10.28: STATUSTEXT {severity : 6, text : 89ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789, id : 217, chunk_num : 4}
```

**Comment by peterbarker** [comment]
The concept:
 - be able to reassemble longer logical statustexts from smaller statustext fragments
 - make it backwards-compatible with mavlink1 / non-extension-aware GCSs
 - two new fields, a statustext id field `uint16_t` and a chunk-offset number `uint8_t`
  - the statustext id field is an opaque number used for reassembly.
  - the chunk-offset-number is an index from zero.  You can multiply this number by 50 to get the offset to place the chunk in the larger statustext field
 - Any statustext message of total length <= 50 characters gets an ID of zero (and a chunk number of 0).
  - these can be flushed straight away
  - maximally backwards compatible
  - allows for mavlink2 zero-trimming for efficiency
 - all *other* strings (longer than 50 characters) are NON-OPTIONALLY null-terminated
 - so a 100 byte string will require a third chunk with zero bytes in the text field (remembering that the text field is 50 bytes in length):

```
2019-12-12 11:15:05.83: STATUSTEXT {severity : 6, text : 99, id : 0, chunk_num : 0}
2019-12-12 11:15:05.84: STATUSTEXT {severity : 6, text : 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01, id : 49, chunk_num : 0}
2019-12-12 11:15:05.84: STATUSTEXT {severity : 6, text : 23456789ABCDEF0123456789ABCDEF0123456789ABCDEF012, id : 49, chunk_num : 1}
2019-12-12 11:15:05.96: STATUSTEXT {severity : 6, text : 100, id : 0, chunk_num : 0}
2019-12-12 11:15:05.97: STATUSTEXT {severity : 6, text : 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01, id : 50, chunk_num : 0}
2019-12-12 11:15:05.97: STATUSTEXT {severity : 6, text : 23456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123, id : 50, chunk_num : 1}
2019-12-12 11:15:05.97: STATUSTEXT {severity : 6, text : , id : 50, chunk_num : 2}
2019-12-12 11:15:06.06: STATUSTEXT {severity : 6, text : 101, id : 0, chunk_num : 0}
2019-12-12 11:15:06.07: STATUSTEXT {severity : 6, text : 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01, id : 51, chunk_num : 0}
2019-12-12 11:15:06.07: STATUSTEXT {severity : 6, text : 23456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123, id : 51, chunk_num : 1}
2019-12-12 11:15:06.07: STATUSTEXT {severity : 6, text : 4, id : 51, chunk_num : 2}
```
 - some sort of timeout will be required in the case that chunks are lost
 - you know when you have received the final chunk in a multi-message statustext because its length is < 50

**Comment by peterbarker** [comment]
MAVProxy output when this is supported:
```
APM: 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789AB
APM: 109
APM: 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABC
APM: 110
APM: 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCD
APM: 111
APM: 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDE
```

MAVProxy output when this is not supported:
```
APM: 109
APM: 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01
APM: 23456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123
APM: 456789ABC
APM: 110
APM: 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01
APM: 23456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123
APM: 456789ABCD
APM: 111
APM: 0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF01
APM: 23456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123
APM: 456789ABCDE
```

File: libraries/GCS_MAVLink/GCS.cpp [hunk 1]
```diff
@@ -37,7 +37,7 @@ const MAV_MISSION_TYPE GCS_MAVLINK::supported_mission_types[] = {
  */
 void GCS::send_textv(MAV_SEVERITY severity, const char *fmt, va_list arg_list)
 {
-    char text[MAVLINK_MSG_STATUSTEXT_FIELD_TEXT_LEN+1];
+    char text[256+1]; // eep!  Stack!  Eeep!
```
Comment (tridge) [inline libraries/GCS_MAVLink/GCS.cpp]: as send_text() can be called from any thread, I think this is too much stack to use

Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.cpp]: Fixed.  We now use a buffer shared buffer.

**Comment by davidbuzz** [comment]
reviewed, looks good to me.      LGTM, and also tell other GCS implementers, as I'm sure MP and QGC will want to support this at some point.  @meee1  and @DonLakeFlyer

**Comment by auturgy** [comment]
I’m pretty sure @peterbarker is the admin for the gcs developers mailing list....

**Comment by DonLakeFlyer** [comment]
Two things I don't understand:
* Why the continued focus on mavlink 1? Mavlink 2 has been out forever.
* This is an ArduPilot mavlink dialect specific change? That's not super helpful.

**Comment by auturgy** [comment]
Don makes valid points. STATUSTEXT_LONG already exists, and no justification is made for not using it.
History on STATUSTEXT_LONG here: https://github.com/mavlink/mavlink/pull/1038
and https://github.com/mavlink/mavlink/pull/1058

This type of thing really makes the mavlink coordination efforts pointless.

**Comment by peterbarker** [comment]
On Sat, 18 Jan 2020, Don Gagne wrote:

>  *  Why the continued focus on mavlink 1? Mavlink 2 has been out forever.

It has, but I continue to see reports from people not running links 
capable of supporting mavlink1.

We are considering changing the default in ArduPilot to mavlink2 soon.

Upping and changing to using STATUSTEXT_LONG is a majorly disruptive 
change.

>  *  This is an ArduPilot mavlink dialect specific change? That's not super helpful.

No, it would definitely go upstream.  Naturally for development it went 
into my fork of ArduPilot's mavlink and that branch pushed into the 
ArduPilot repo so our CI can pass.  It did change shape during 
development.

I do intend to create the PR upstream if @amilcarlucas doesn't do it 
first.

**Comment by WickedShell** [comment]
@DonLakeFlyer cause STATUSTEXT_LONG still sticks you with a much smaller maximum message size, and you then have the problem of having to send both until you've given GCS's sufficent time to swap their implementations. The proposed one here just works (albeit not in the best  mode) without requiring a GCS to be updated.

**Comment by peterbarker** [comment]
Merged after discussions with tridge and MdB

**Pull request merged**